### PR TITLE
[Homematic] Fix "Channel not found for Datapoint"-Errors

### PR DESCRIPTION
**Pull request by Flole998**
Issue described and tested in https://community.openhab.org/t/homematic-binding-channel-not-found-for-datapoint-errors-for-definitely-existing-channels/26209/113 

Fixes: #11214
Signed-off-by: Flole <flole@flole.de>

### CODE REVIEW

**Comment by Flole998** [comment]
@FStolte @gerrieg @mdicke2s Can anybody review please?

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java [hunk 1]
```diff
@@ -150,6 +150,42 @@ private void doInitializeInBackground() throws GatewayNotAvailableException, Hom
         // update thing channel list for reconfigurable channels (relies on the new value of the
         // CHANNEL_FUNCTION datapoint fetched during configuration update)
         List<Channel> thingChannels = new ArrayList<>(getThing().getChannels());
+
+        if (thingChannels.size() == 0) {
```
Comment (maniac103) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:154]: Using `isEmpty()` seems more clear.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java [hunk 2]
```diff
@@ -150,6 +150,42 @@ private void doInitializeInBackground() throws GatewayNotAvailableException, Hom
...
+                            "Updated value datapoints for channel {} of device '{}' (function {}), now has {} datapoints",
+                            channel, channel.getDevice().getAddress(), channel.getCurrentFunction(),
+                            channel.getDatapoints().size());
+                }
+            }
+            updateThing(editThing().withChannels(thingChannels).build());
+        }
+
         if (updateDynamicChannelList(device, thingChannels)) {
             updateThing(editThing().withChannels(thingChannels).build());
```
Comment (maniac103) [inline bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:190]: Should collapse both `updateThing()` calls into one, e.g. by having a boolean variable `thingListChanged` which is set to `true` in your new block and updated here as well (`changed |= updateDynamicChannelList(...)`), and then `updateThing` is called when `changed` is `true`.

**Comment by maniac103** [comment]
What still irks me about this is that this change is just a workaround for what should be a bug elsewhere. Did you try adding some breakpoints in `ThingFactory` or `BaseThingHandlerFactory` to check the state of the `ThingType` at thing creation time? 
AFAICT, the flow is
- inbox contains thing UID and thing type UID
- inbox calls `ThingFactory.createThing()`, passing thing UID, thing type UID, and bridge UID
- `ThingFactory.createThing()` tries to lookup a `ThingHandlerFactory` based on the thing type UID; if this fails, no thing is created
- `ThingFactory.craeteThing()` calls `createThing()` in the found factory (here, it's `HomematicThingHandlerFactory`)
- `HomematicThingHandlerFactory.createThing` (or rather the base class implementation in `BaseThingHandlerFactory`) looks up the `ThingType` by the thing type UID via `ThingTypeRegistry`, which in turn looks through all registered `ThingTypeProvider`s. In our case, `HomematicThingTypeProviderImpl` should match. If no `ThingType` can be found, no thing is created.
- `ThingFactory.createThing` is called, passing in the found `ThingType`, which in turn calls `ThingFactoryHelper.createChannels`
- `ThingFactoryHelper.createChannels` loops over `thingType.getChannelDefinitions()`, calling `ThingFactoryHelper.createChannel` for each one, which in turn calls `ThingFactoryHelper.createChannelBuilder`
- That one looks up a `ChannelType` from the channel type UID in the channel definition using a call to `ChannelTypeRegistry.getChannelType()`
- That one looks through all `ChannelTypeProvider`s (here: `HomematicChannelTypeProviderImpl` for a matching channel type

AFAICT, the *only* case where a thing can be created without channels is if the channel type lookup fails, so it would be great if you could verify this actually being the case. If it is, we need to check what's off in channel type registration. Unfortunately it's hard to follow git history due to the code being moved between repositories several times.

**Comment by Flole998** [comment]
The reason this is necessary is, that (at least) when using text based config it could happen that things are created before the bridge is online or the thing type has been populated with the channels. In that case an empty thing type is used and the thing is missing the channels as when a thing type is updated the things that use it aren't automatically. So this will add the channels in that case.

**Comment by maniac103** [comment]
So this essentially means that `createThing()` is called before `generate()` is called, right?

**Comment by Flole998** [comment]
Yes iirc that was the issue. I also tried to re-create (basically reload) the thing if the channels are empty and that created the channels as well (but caused other issues).

If you ask me the current attempt is a wrong approach anyways as there's another Limit: If there are multiple versions of the same device with different channels it will cause issues. It would be better to request the channels for each thing and populate them on initialization.

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java (hunk 1 above)
Comment (Flole998) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java]: Fixed

File: bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java (hunk 2 above)
Comment (Flole998) [reply bundles/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/handler/HomematicThingHandler.java:196]: Fixed

**Comment by Flole998** [comment]
@maniac103 I just fixed the 2 things you commented.

**Review by fwolter — APPROVED** [review]

**Comment by fwolter** [comment]
The sign-off message should contain your real name. For the next contribution...
